### ISSUE TITLE
Header issue PHP_AUTH_PW

### ISSUE THREAD

**Issue by bkcummins** [bshaffer/oauth2-server-httpfoundation-bridge#21]
First off thx for the great work done here!

I've used the oauth-server-php library in Laravel to build out an oauth api for our microservices platform with a separate angular client to manage clients, keys, users, scope, etc.

I've worked through all the grant type workflows successfully; however, in an attempt to send client_id and client_secret (when needed) through as headers (PHP-AUTH-USER and PHP-AUTH-PW, respectively) instead of in the post body I'm running into issues.  I'm using nginx to proxy php5-fpm with Laravel 5.

The latest thing I encountered was the BaseRequest which uses Symfony\Component\HttpFoundation\Request here:
https://github.com/bshaffer/oauth2-server-httpfoundation-bridge/blob/develop/src/OAuth2/HttpFoundationBridge/Request.php#L39

Their getHeaders() methods throws an error when you use the password grant type for a public client.

I made a pull request to them here stating what I'm running into.
https://github.com/symfony/http-foundation/pull/29

I tried extending OAuth2\HttpFoundationBridge\Request and Symfony\Component\HttpFoundation\Request but got incompatibility exception for the BaseRequest.

Curious if you have a better way to approach this.  Thx again!

**Comment by bshaffer** [comment]
Can you show an example of your request? What line of code is the error being thrown on? What happens if you use an empty string for the password?

**Comment by bkcummins** [comment]
Thx...just to recap for clarity sake.

This a workflow for an angular login client to obtain a jwt access_token used for resource servers in a micro services environment behind the same FQDN using a nginx reverse proxy....so grant_type is password, no client_secret since it's JS. 

I haven't found a way to add the grant_type to the header so grant_type, username and password are in the post body.  "client_id" didn't register as a header but I saw with the Symfony ServerBag getHeaders() method you could add it as a header with PHP-AUTH-USER. 

Below is a call to the token controller and the subsequent error:
Undefined index: PHP_AUTH_PW' in /oauth/vendor/symfony/http-foundation/ServerBag.php:95   

<img width="1024" alt="screen shot 2016-05-06 at 09 51 03" src="https://cloud.githubusercontent.com/assets/6819212/15074987/1c929578-1370-11e6-9348-73dd4b54c9a2.png">

![image](https://cloud.githubusercontent.com/assets/6819212/15074995/2509fd9a-1370-11e6-8b71-ac5b84949948.png)

**Comment by bkcummins** [comment]
Here's a screenshot to show it works in post body (headers remove) with or without client_secret submitted.

![image](https://cloud.githubusercontent.com/assets/6819212/15075196/19062770-1371-11e6-90f8-6c5e9f5e5a6e.png)

![image](https://cloud.githubusercontent.com/assets/6819212/15075206/23a45954-1371-11e6-9899-133019fb39db.png)

**Comment by bkcummins** [comment]
Of course all this begs the question which is better practice for sending client auth parameters...in the header or body.  Is it best just to use post body?  Looking at the RFC OAuth 2.0 specs 2.3.1. for Client Password the example shows client_id and client_secret in post body.

![image](https://cloud.githubusercontent.com/assets/6819212/15075600/41c5972a-1373-11e6-9aa1-27ea5f50a9a4.png)

https://tools.ietf.org/html/rfc6749#section-2.3.1

**Comment by steffkes** [comment]
@bkcummins http-header are typically all uppercase and **underscore** rather than dashes, which might explain as well why Symfonys `ParameterBag` is complaining about a missing `PHP_AUTH_PW`

**Comment by bshaffer** [comment]
What web server are you using? I am assuming it's not setting `PHP_AUTH_PW`, and you'll need to add a rewrite rule in `.htaccess`:

```
RewriteEngine on
RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
```

Also make sure [`fixAuthHeaders`](https://github.com/TrogloGeek/oauth2-server-httpfoundation-bridge/blob/d10366e8da92479ea63ab693667c3ef0d8936792/src/OAuth2/HttpFoundationBridge/Request.php#L58) is being called

**Comment by bkcummins** [comment]
@bshaffer This is all running on nginx. There's a nginx reverse proxy orchestration node in the front which pulls from a nginx/php-fpm node running laravel 5.  

Passing Authorization as a header worked fine. The issue above is to obtain an access_token via user_crendentials public client workflow....when there's a client_id and no client_secret or PHP_AUTH_USER and PHP_AUTH_PW, respectively, if passed as http headers.

@steffkes I had no problem getting Symfony to recognize PHP-AUTH-USER and PHP-AUTH-PW...and everything works when you don't use the public client workflow.  It's when PHP_AUTH_PW is not passed or even attempting to pass it empty that there's an error.

In case anyone find this helpful...here's a link to nginx docs about underscores in headers, but I found when I passed PHP-AUTH-USER as a header it was actually translated to PHP_AUTH_USER.  I didn't trace whether that was a Symfony function or in the proxy layer.

underscores_in_headers
http://nginx.org/en/docs/http/ngx_http_core_module.html#underscores_in_headers

I was interested in trying to get it work with http headers; however, in the end I stuck with the format shown in the spec up just sticking with the RFC specs listed above.

As far as I know so far both approaches are equally safe give proper TLS procedures and as long as "secret" variables aren't logged in an insecure manner.

Let me know if I'm off on any of this....I think as an issue for this repo it could close, but we could consider some clarification on when to use http headers vs post body in the documentation.

Thx again for all your hard work.  bkc
